pkg/conversion/capi2mapi/powervs.go

```diff
@@ -0,0 +1,305 @@
+		Image:           image,
+		Network:         network,
+		KeyPairName:     m.powerVSMachine.Spec.SSHKey,
+		SystemType:      m.powerVSMachine.Spec.SystemType,
+		ProcessorType:   mapiv1.PowerVSProcessorType(m.powerVSMachine.Spec.ProcessorType),
+		Processors:      m.powerVSMachine.Spec.Processors,
+		MemoryGiB:       m.powerVSMachine.Spec.MemoryGiB,
+		//CredentialsSecret:
+		//LoadBalancers: TODO: Not supported for workers.
+	}
+
+	userDataSecretName := ptr.Deref(m.machine.Spec.Bootstrap.DataSecretName, "")
+	if userDataSecretName != "" {
+		mapiProviderConfig.UserDataSecret = &mapiv1.PowerVSSecretReference{
+			Name: userDataSecretName,
+		}
+	}
+
+	return &mapiProviderConfig, warnings, nil
+}
+
+// Conversion helpers.
+
+// TODO: May be we can use generics and support for all the platforms?
+// rawExtensionFromPowerVSProviderSpec marshals the machine provider spec.
+func rawExtensionFromPowerVSProviderSpec(spec *mapiv1.PowerVSMachineProviderConfig) (*runtime.RawExtension, error) {
+	if spec == nil {
+		return &runtime.RawExtension{}, nil
+	}
+
+	rawBytes, err := json.Marshal(spec)
+	if err != nil {
+		return nil, fmt.Errorf("error marshalling providerSpec: %w", err)
+	}
+
+	return &runtime.RawExtension{
+		Raw: rawBytes,
+	}, nil
+}
+
+func convertPowerVSNetworkToMAPI(fldPath *field.Path, network capibmv1.IBMPowerVSResourceReference) (mapiv1.PowerVSResource, *field.Error) {
+	var networkResource mapiv1.PowerVSResource
+	if network.ID != nil {
+		networkResource.Type = mapiv1.PowerVSResourceTypeID
+		networkResource.ID = network.ID
+		return networkResource, nil
+	}
+	if network.Name != nil {
+		networkResource.Type = mapiv1.PowerVSResourceTypeName
+		networkResource.Name = network.Name
+		return networkResource, nil
+	}
+	if network.RegEx != nil {
+		networkResource.Type = mapiv1.PowerVSResourceTypeRegEx
+		networkResource.RegEx = network.RegEx
+		return networkResource, nil
+	}
+	return networkResource, field.Invalid(fldPath, network, "unable to convert network to MAPI")
+}
+
+func convertPowerVSImageToMAPI(fldPath *field.Path, image *capibmv1.IBMPowerVSResourceReference, imageRef *corev1.LocalObjectReference) (mapiv1.PowerVSResource, *field.Error) {
+	if image == nil && imageRef == nil {
+		return mapiv1.PowerVSResource{}, field.Invalid(fldPath, image, "unable to convert image, image and imageref is nil")
+	}
+	var imageResource mapiv1.PowerVSResource
+	if image != nil {
+		if image.ID != nil {
+			imageResource.Type = mapiv1.PowerVSResourceTypeID
+			imageResource.ID = image.ID
+			return imageResource, nil
+		}
+		if image.Name != nil {
+			imageResource.Type = mapiv1.PowerVSResourceTypeName
+			imageResource.Name = image.Name
+			return imageResource, nil
+		}
+		if image.RegEx != nil {
+			imageResource.Type = mapiv1.PowerVSResourceTypeRegEx
+			imageResource.RegEx = image.RegEx
+			return imageResource, nil
+		}
+	}
+	imageResource.Type = mapiv1.PowerVSResourceTypeName
+	imageResource.Name = &imageRef.Name
```

Comment:
Wonder if refactoring this to a switch might make things simpler? 

I assume again the ordering here is similar to the implementation? You have no discriminator on the MAPI side right?

Reply:
updated it to use switch.